### DISCUSSION TITLE
FormItem 提供一个新的属性来决定是否让表单 Form 收集

### DISCUSSION THREAD

**Discussion by guaijie** [ant-design/ant-design#46466] [Q&A]
通常来讲，readonly 和 disabled 表现应该是不一致的，form 表单会收集 readonly 的表单项数据，而不会收集 disabled 表单项数据。现阶段，我不知道如何做才能不去收集指定的字段。是否有相应的属性可以传递给 FormItem 或 表单组件，使得 Form 不去收集该表单项数据

**Comment by zombieJ** [comment]
感觉可以 HOC 一下 Form.Item，如果遇到 `disabled` 直接把 `name` 干掉，通过 `getFieldValue` 赋值。

**Reply by guaijie** [reply]
我还是建议区分 disabled 和 readonly 或者新增一个 prop 传递给 FormItem

**Reply by yoyo837** [reply]
单一职责？

**Reply by guaijie** [reply]
优先建议区分 disabled 和 readonly，我印象中宿主环境中的 input 元素如果使用 disabled 是不会收集该数据的。
我们可以使 disabled 和 readonly 样式表现一致但功能不一样

**Reply by guaijie** [reply]
使用 getFieldValue  给表单赋值是个无奈之举，这意味着我得分别单独处理。而不能通过 form.setFieldValues 或者 initialValues 统一赋值。但我可以通过 validateFields 和 getFieldValues 只拿到没有被 disabled 的数据

**Reply by crazyair** [reply]
按理说应该是 只读的，那可以只用 div 展示，表单就不会有这个字段，而禁用的，应该用表单展示(可能根据条件判断是否禁用)，应该有表单字段

**Reply by guaijie** [reply]
这样做的目的只是更好的语义化和更简单的控制，而不是用什么 div 展示。我要实现我所表达的有很多种方案，问题是要提供一种与宿主环境一致的，更为语义化和更好控制的方案。
form 表单管理这个字段和是否收集这个字段是两回事。谁说这个字段被 form 管理了就一定要去收集。可以默认收集，但并不意味着一定要收集，我可以通过其他属性判断。但通常，为了更好的语义化表述，同时与要被收集的字段进行区分，所以才说推荐用 hidden。
这就意味着，这些被禁用的字段需要被 form 管理，同时也要不被收集。如果为了兼容之前的版本，可以推荐先用一个新的属性来控制。

**Reply by guaijie** [reply]
我更为推荐用 readonly 代替现在的 hidden，如果可以话请将其作为一个 feature request
